agent/pkg/handler/linkerd_info.go

```diff
@@ -75,3 +92,53 @@ func (h *LinkerdInfo) handleCertsInfo() {
 		h.log.Errorf("error sending CertificateInfo message: %s", err)
 	}
 }
+
+func (h *LinkerdInfo) handleProxyDiagnostics(podName string, namespace string, diagnosticsId string) {
+	initialLogLevel, err := h.k8s.GetProxyLogLevel(podName, namespace)
+	if err != nil {
+		h.log.Errorf("error getting proxy log level for pod %s/%s: %s", namespace, podName, err)
+		return
+	}
+
+	err = h.k8s.SetProxyLogLevel(podName, namespace, "trace")
+	if err != nil {
+		h.log.Errorf("error setting proxy log level for pod %s/%s: %s", namespace, podName, err)
+		return
+	}
+
+	time.AfterFunc(diagnosticCollectDuration, func() {
+		logs, err := h.k8s.GetProxyLogs(podName, namespace)
+		if err != nil {
+			h.log.Errorf("cannot obtain proxy logs for pod %s/%s: %s", namespace, podName, err)
+			return
+		}
+
+		metrics, err := h.k8s.GetPrometheusScrape(podName, namespace)
+		if err != nil {
+			h.log.Errorf("cannot obtain proxy metrics for pod %s/%s: %s", namespace, podName, err)
+			return
+		}
+
+		data, err := yaml.Marshal(DiagnosticsData{Logs: logs, Metrics: metrics})
+		if err != nil {
+			h.log.Errorf("cannot marshal diagnostic data for pod %s/%s: %s", namespace, podName, err)
+			return
+		}
+		m := &pb.ProxyDiagnostics{
+			DiagnosticId: diagnosticsId,
+			Data:         data,
+		}
+		h.log.Tracef("handleLinkerdInfo: %s", prototext.Format(m))
+
+		err = h.api.ProxyDiagnostics(m)
+		if err != nil {
+			h.log.Errorf("error sending ProxyDiagnostics message: %s", err)
+		}
+
+		err = h.k8s.SetProxyLogLevel(podName, namespace, initialLogLevel)
+		if err != nil {
+			h.log.Errorf("error reverting proxy log level to %s for pod %s/%s: %s", initialLogLevel, namespace, podName, err)
+			return
+		}
```

Comment:
i think we'd want this in a `defer` to ensure it always gets called.

going a step further, related to my comment around separating log level and diagnostics collection, this may be moot.

Reply:
Good point